File: extension/src/parser.ts
/**
 * Catapillar parser — TypeScript port of parser/parser.py.
 * Converts tokens into an AST.
 */

import { Token, tokenizeSource, TokenizeWarning } from './tokenizer';
import { ACTION_LOOKUP, STRUCT_LOOKUP, LINE_STATE_SET } from './keywords';

// =====================================================
// AST Node Types
// =====================================================

export interface ProgramNode {
    type: 'Program';
    flows: FlowNode[];
}

export interface FlowNode {
    type: 'Flow';
    segments: SegmentNode[];
}

export interface SegmentNode {
    type: 'Segment';
    lines: StatementNode[];
}

export type StatementNode = LineNode | BlockNode | ArrowNode | BlockEndNode;

export interface LineNode {
    type: 'Line';
    action: string;
    args: string[];
    lineState: string;
    lineno: number;
}

export interface BlockNode {
    type: 'Block';
    name: string;
    lineState: string;
    lineno: number;
}

export interface ArrowNode {
    type: 'Arrow';
    from: string;
    to: string;
    direction: '->' | '<-';
    lineState: string;
    lineno: number;
}

export interface BlockEndNode {
    type: 'BLOCK_END';
    lineState: string;
    lineno: number;
}

// =====================================================
// Parse Errors & Diagnostics
// =====================================================

export interface ParseDiagnostic {
    message: string;
    lineno: number;
    severity: 'error' | 'warning';
    colStart?: number;
    colEnd?: number;
}

// =====================================================
// Parser
// =====================================================

export interface ParseResult {
    program: ProgramNode;
    diagnostics: ParseDiagnostic[];
    symbols: SymbolEntry[];
}

export interface SymbolEntry {
    name: string;
    kind: 'function' | 'variable' | 'block';
    lineno: number;
    params?: string[];
}

/**
 * Parse source code into a Catapillar AST with diagnostics.
 * Unlike the Python parser, this does NOT throw on errors — it collects
 * diagnostics so the extension can show them all at once.
 */
export function parseSource(source: string): ParseResult {
    const { tokens, warnings } = tokenizeSource(source);

    const diagnostics: ParseDiagnostic[] = warnings.map(w => ({
        message: w.message,
        lineno: w.lineno,
        severity: 'warning' as const,
    }));

    const symbols: SymbolEntry[] = [];

    const program: ProgramNode = {
        type: 'Program',
        flows: [],
    };

    const currentFlow: FlowNode = {
        type: 'Flow',
        segments: [],
    };

    let currentSegment: SegmentNode = {
        type: 'Segment',
        lines: [],
    };

    const blockStack: string[] = [];

    for (const token of tokens) {
        const { rawAction, rawArgs, lineState, lineno } = token;

        // Struct keywords (block end)
        if (STRUCT_LOOKUP.has(rawAction)) {
            const structId = STRUCT_LOOKUP.get(rawAction)!;
            if (structId === 'BLOCK_END') {
                if (blockStack.length === 0) {
                    diagnostics.push({
                        message: `Unexpected block end '${rawAction}' without matching block opener`,
                        lineno,
                        severity: 'warning',
                    });
                } else {
                    blockStack.pop();
                }
                currentSegment.lines.push({
                    type: 'BLOCK_END',
                    lineState,
                    lineno,
                });
            }
            continue;
        }

        // Validate line state
        if (!LINE_STATE_SET.has(lineState)) {
            diagnostics.push({
                message: `Invalid line state: ${lineState}`,
                lineno,
                severity: 'error',
            });
            continue;
        }

        // Detect arrow syntax
        let arrowType: '->' | '<-' | null = null;
        if (rawArgs.includes('->')) { arrowType = '->'; }
        else if (rawArgs.includes('<-')) { arrowType = '<-'; }

        // Block definition (action ends with :)
        if (rawAction.endsWith(':')) {
            const blockName = rawAction.slice(0, -1);

            // Check for control keywords that end with colon
            if (['否则', 'else'].includes(blockName)) {
                // ELSE is continuation of IF/ELIF chain — pop previous branch
                if (blockStack.length > 0 && (blockStack[blockStack.length - 1] === 'IF' || blockStack[blockStack.length - 1] === 'ELIF')) {
                    blockStack.pop();
                }
                blockStack.push('ELSE');
                currentSegment.lines.push({
                    type: 'Line',
                    action: 'ELSE',
                    args: [],
                    lineState,
                    lineno,
                });
                continue;
            }
            if (['试', 'try'].includes(blockName)) {
                blockStack.push('TRY');
                currentSegment.lines.push({
                    type: 'Line',
                    action: 'TRY',
                    args: [],
                    lineState,
                    lineno,
                });
                continue;
            }
            if (['终于', 'finally'].includes(blockName)) {
                // FINALLY is continuation of TRY/EXCEPT chain — pop previous branch
                if (blockStack.length > 0 && (blockStack[blockStack.length - 1] === 'TRY' || blockStack[blockStack.length - 1] === 'EXCEPT')) {
                    blockStack.pop();
                }
                blockStack.push('FINALLY');
                currentSegment.lines.push({
                    type: 'Line',
                    action: 'FINALLY',
                    args: [],
                    lineState,
                    lineno,
                });
                continue;
            }

            // Named flow block (e.g. ~ 成功流程:) — push so matching 'end' pops
            blockStack.push('FLOW_BLOCK');
            currentSegment.lines.push({
                type: 'Block',
                name: blockName,
                lineState,
                lineno,
            });
            continue;
        }

        // Arrow lines
        if (arrowType) {
            const idx = rawArgs.indexOf(arrowType);
            if (idx + 1 >= rawArgs.length) {
                diagnostics.push({
                    message: 'Arrow missing target',
                    lineno,
                    severity: 'error',
                });
                continue;
            }

            const left = rawAction;
            const right = rawArgs[idx + 1];

            currentSegment.lines.push({
                type: 'Arrow',
                from: arrowType === '->' ? left : right,
                to: arrowType === '->' ? right : left,
                direction: arrowType,
                lineState,
                lineno,
            });
        }
        // Legacy action lines
        else {
            // Strip trailing colon from action for lookup
            let lookupAction = rawAction;
            let processedArgs = [...rawArgs];
            if (lookupAction.endsWith(':')) {
                lookupAction = lookupAction.slice(0, -1);
            }

            if (!ACTION_LOOKUP.has(lookupAction)) {
                diagnostics.push({
                    message: `Unknown action: ${rawAction}`,
                    lineno,
                    severity: 'error',
                    colStart: token.colOffset,
                    colEnd: token.colOffset + rawAction.length,
                });
                continue;
            }

            const actionId = ACTION_LOOKUP.get(lookupAction)!;

            // Track block-opening actions (ELIF/EXCEPT replace previous branch on stack)
            if (actionId === 'ELIF') {
                if (blockStack.length > 0 && (blockStack[blockStack.length - 1] === 'IF' || blockStack[blockStack.length - 1] === 'ELIF')) {
                    blockStack.pop();
                }
                blockStack.push(actionId);
            } else if (actionId === 'EXCEPT') {
                if (blockStack.length > 0 && (blockStack[blockStack.length - 1] === 'TRY' || blockStack[blockStack.length - 1] === 'EXCEPT')) {
                    blockStack.pop();
                }
                blockStack.push(actionId);
            } else if (['DEF', 'IF', 'WHILE', 'FOR', 'TRY'].includes(actionId)) {
                blockStack.push(actionId);
            }

            // Collect symbol info for definitions
            if (actionId === 'DEF' && rawArgs.length > 0) {
                const funcName = rawArgs[0].replace(/:$/, '');
                const params = rawArgs.slice(1).map(p => p.replace(/:$/, ''));
                symbols.push({
                    name: funcName,
                    kind: 'function',
                    lineno,
                    params,
                });
            }
            if (actionId === 'SET' && rawArgs.length >= 2) {
                symbols.push({
                    name: rawArgs[0],
                    kind: 'variable',
                    lineno,
                });
            }

            // Strip trailing colon from last arg (e.g. "若 cond:" → args with colon)
            if (processedArgs.length > 0 && processedArgs[processedArgs.length - 1].endsWith(':')) {
                processedArgs[processedArgs.length - 1] =
                    processedArgs[processedArgs.length - 1].slice(0, -1);
            }

            currentSegment.lines.push({
                type: 'Line',
                action: actionId,
                args: processedArgs,
                lineState,
                lineno,
            });
        }

        // Segment closing on >
        if (lineState === '>') {
            currentFlow.segments.push(currentSegment);
            currentSegment = { type: 'Segment', lines: [] };
        }
    }

    // Check for unclosed blocks
    if (blockStack.length > 0) {
        diagnostics.push({
            message: `Unclosed block(s): expected ${blockStack.length} more '终/end' statement(s)`,
            lineno: tokens.length > 0 ? tokens[tokens.length - 1].lineno : 1,
            severity: 'warning',
        });
    }

    // Flush
    if (currentSegment.lines.length > 0) {
        currentFlow.segments.push(currentSegment);
    }
    if (currentFlow.segments.length > 0) {
        program.flows.push(currentFlow);
    }

    return { program, diagnostics, symbols };
}

/**
 * Get all defined function names from a parsed result.
 */
export function getFunctionSymbols(result: ParseResult): SymbolEntry[] {
    return result.symbols.filter(s => s.kind === 'function');
}

/**
 * Get all variable assignments from a parsed result.
 */
export function getVariableSymbols(result: ParseResult): SymbolEntry[] {
    return result.symbols.filter(s => s.kind === 'variable');
}

File: plugin/runtime/parser/parser.py
# parser/parser.py
# Catapillar parser: token list → AST.
# Supports arrow lines and legacy action lines. Expression parsing lives in parser.expr.

from typing import List, Dict

from parser.errors import ParseError
from parser.tokenizer import tokenize_expression
from parser.constants import ACTION_LOOKUP, STRUCT_LOOKUP, LINE_STATES, IN_KEYWORDS
from parser.expr import (
    parse_expression,
    parse_condition,
    parse_set_value,
    parse_print_value,
    try_parse_expr_full,
    strip_trailing_colon,
)


# ============================================================
# Core Parser
# ============================================================

def _declared_names(action_id: str, line_node: Dict, raw_args: List[str]) -> set:
    """Names declared by this line (for print variable resolution)."""
    out = set()
    if action_id == "SET" and "name" in line_node:
        out.add(line_node["name"])
    elif action_id == "INDEX_SET" and "container" in line_node:
        out.add(line_node["container"])
    elif action_id == "FOR" and "var" in line_node:
        out.add(line_node["var"])
    elif action_id == "DEF" and raw_args:
        name = raw_args[0].rstrip(":")
        if name:
            out.add(name)
        for a in raw_args[1:]:
            s = a.rstrip(":")
            if s:
                out.add(s)
    return out


def parse_tokens(tokens: List[Dict]) -> Dict:
    """
    Convert token list into Catapillar AST.
    Supports arrow lines (intent flow) and legacy action lines (with expression parsing).
    Tracks known_names (variables declared so far) so 印 can resolve variable vs string.
    """
    program = {"type": "Program", "flows": []}
    current_flow = {"type": "Flow", "segments": []}
    current_segment = {"type": "Segment", "lines": []}
    known_names = set()

    for token in tokens:
        raw_action = token["raw_action"]
        raw_args = token["raw_args"]
        line_state = token["line_state"]
        emoji = token.get("emoji_prefix")

        if raw_action in STRUCT_LOOKUP:
            struct_id = STRUCT_LOOKUP[raw_action]
            if struct_id == "BLOCK_END":
                line_node = {"type": "BLOCK_END", "line_state": line_state}
                if emoji:
                    line_node["emoji"] = emoji
                current_segment["lines"].append(line_node)
                continue

        if line_state not in LINE_STATES:
            raise ParseError(f"Invalid line_state: {line_state}")

        arrow_type = None
        if "->" in raw_args:
            arrow_type = "->"
        elif "<-" in raw_args:
            arrow_type = "<-"

        if raw_action.endswith(":"):
            block_name = raw_action[:-1]
            if block_name in ["否则", "else"]:
                line_node = {"type": "Line", "action": "ELSE", "args": [], "line_state": line_state}
                if emoji:
                    line_node["emoji"] = emoji
                current_segment["lines"].append(line_node)
                continue
            elif block_name in ["试", "try"]:
                line_node = {"type": "Line", "action": "TRY", "args": [], "line_state": line_state}
                if emoji:
                    line_node["emoji"] = emoji
                current_segment["lines"].append(line_node)
                continue
            elif block_name in ["终于", "finally"]:
                line_node = {"type": "Line", "action": "FINALLY", "args": [], "line_state": line_state}
                if emoji:
                    line_node["emoji"] = emoji
                current_segment["lines"].append(line_node)
                continue
            else:
                block_node = {"type": "Block", "name": block_name, "lines": [], "line_state": line_state}
                current_segment["lines"].append(block_node)
                known_names.add(block_name)
                continue

        if arrow_type:
            idx = raw_args.index(arrow_type)
            if idx + 1 >= len(raw_args):
                raise ParseError("Arrow missing target")
            left, right = raw_action, raw_args[idx + 1]
            from_node, to_node = (left, right) if arrow_type == "->" else (right, left)
            line_node = {"type": "Arrow", "from": from_node, "to": to_node, "direction": arrow_type, "line_state": line_state}
            current_segment["lines"].append(line_node)
        else:
            if raw_action not in ACTION_LOOKUP:
                raise ParseError(f"Unknown action: {raw_action}")
            action_id = ACTION_LOOKUP[raw_action]
            line_node = _build_action_line(action_id, raw_args, line_state, emoji, known_names)
            current_segment["lines"].append(line_node)
            known_names |= _declared_names(action_id, line_node, raw_args)

        if line_state == ">":
            current_flow["segments"].append(current_segment)
            current_segment = {"type": "Segment", "lines": []}

    if current_segment["lines"]:
        current_flow["segments"].append(current_segment)
    if current_flow["segments"]:
        program["flows"].append(current_flow)

    return program


# ============================================================
# Action Line Builder
# ============================================================

def _build_action_line(
    action_id: str, raw_args: List[str], line_state: str, emoji: str, known_names: set = None
) -> Dict:
    if known_names is None:
        known_names = set()
    base = {"type": "Line", "action": action_id, "line_state": line_state}
    if emoji:
        base["emoji"] = emoji

    if action_id == "SET":
        if not raw_args:
            raise ParseError("SET expects arguments")
        expr_tokens = tokenize_expression(raw_args)

        if len(expr_tokens) >= 5 and expr_tokens[1] == "[":
            depth, close = 0, None
            for i in range(1, len(expr_tokens)):
                if expr_tokens[i] == "[":
                    depth += 1
                elif expr_tokens[i] == "]":
                    depth -= 1
                    if depth == 0:
                        close = i
                        break
            if close is not None and close + 1 < len(expr_tokens):
                base["action"] = "INDEX_SET"
                base["container"] = expr_tokens[0]
                idx_node, _ = parse_expression(
                    expr_tokens[2:close], 0, value_context=True, known_names=known_names
                )
                base["index"] = idx_node
                base["value"] = parse_set_value(expr_tokens[close + 1:], known_names=known_names)
                base["args"] = raw_args
                return base

        base["name"] = expr_tokens[0]
        base["value"] = (
            parse_set_value(expr_tokens[1:], known_names=known_names)
            if len(expr_tokens) > 1
            else {"type": "NoneLiteral"}
        )
        base["args"] = raw_args
        return base

    if action_id == "PRINT":
        base["value"] = parse_print_value(raw_args, known_names=known_names)
        base["args"] = raw_args
        return base

    if action_id in ("IF", "ELIF", "WHILE"):
        base["condition"] = parse_condition(raw_args, known_names=known_names)
        base["args"] = raw_args
        return base

    if action_id == "FOR":
        if len(raw_args) < 3:
            raise ParseError("FOR expects: variable IN iterable")
        var = raw_args[0]
        if raw_args[1] not in IN_KEYWORDS:
            raise ParseError(f"FOR expects IN keyword, got: {raw_args[1]}")
        iterable_args = strip_trailing_colon(raw_args[2:])
        node, ok = try_parse_expr_full(
            iterable_args, value_context=True, known_names=known_names
        )
        if not ok:
            node = {"type": "Identifier", "name": " ".join(iterable_args)}
        base["var"] = var
        base["iterable"] = node
        base["args"] = raw_args
        return base

    if action_id == "RETURN":
        if raw_args:
            node, ok = try_parse_expr_full(
                raw_args, value_context=True, known_names=known_names
            )
            base["value"] = node if ok else {"type": "Identifier", "name": " ".join(raw_args)}
        else:
            base["value"] = None
        base["args"] = raw_args
        return base

    if action_id == "CALL":
        if not raw_args:
            raise ParseError("CALL expects a function name")
        base["func"] = raw_args[0]
        call_args = []
        for a in raw_args[1:]:
            node, ok = try_parse_expr_full(
                [a], value_context=True, known_names=known_names
            )
            call_args.append(node if ok else {"type": "Identifier", "name": a})
        base["call_args"] = call_args
        base["args"] = raw_args
        return base

    if action_id == "GLOBAL":
        base["names"] = list(raw_args)
        return base

    base["args"] = raw_args
    return base


def parse_file(path: str) -> Dict:
    """Tokenize + parse in one step."""
    from parser.tokenizer import tokenize_file
    tokens = tokenize_file(path)
    return parse_tokens(tokens)
